src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelStateDictionary.cs

```diff
@@ -904,6 +914,224 @@ public ModelStateNode GetOrAddNode(StringSegment subKey)
             public override ModelStateEntry GetModelStateForProperty(string propertyName)
                 => GetNode(new StringSegment(propertyName));
 
+            public override ModelStateEntry GetOrAddModelStateForProperty(
+                ModelStateDictionary dictionary,
+                string propertyName,
+                string key)
+            {
+                if (dictionary == null)
+                {
+                    throw new ArgumentNullException(nameof(dictionary));
+                }
+
+                if (key == null)
+                {
+                    throw new ArgumentNullException(nameof(key));
+                }
+
+                if (string.IsNullOrEmpty(propertyName))
+                {
+                    return this;
+                }
+
+                // Check whether propertyName moves a single level. [Bind(Prefix ="junk.junk[0]")] is completely valid.
+                for (var i = 0; i < propertyName.Length; i++)
+                {
+                    var ch = propertyName[i];
+                    if (ch == DelimiterDot)
+                    {
+                        return dictionary.GetOrAddNode(key);
+                    }
+
+                    if (i != 0 && ch == DelimiterOpen)
+                    {
+                        return dictionary.GetOrAddNode(key);
+                    }
+                }
+
+                return GetOrAddNode(new StringSegment(propertyName));
+            }
+
+            public override ModelValidationState GetFieldValidationState(ModelStateDictionary dictionary, string key)
+            {
+                if (dictionary == null)
+                {
+                    throw new ArgumentNullException(nameof(dictionary));
+                }
+
+                if (key == null)
+                {
+                    throw new ArgumentNullException(nameof(key));
+                }
+
+                return GetValidity(this) ?? ModelValidationState.Unvalidated;
+            }
+
+            public override void SetModelValue(
```

Comment:
Probably some duplicate code in the `ModelStateDictionary` methods can be eliminated using these new methods.